File: script/profiler-gcc.lua

-- use %. to define .
max_funcname_length = 16
exclude_section_prefix ={ '%.text%.', "%.text_nmprofiler" }
-- exclude_section_prefix ={  }
include_section = {  }
exclude_name = { 
"_result_code"                 ,
"loadStackAddr"                ,
"start"                        ,
"_sync"                        ,
"_exit"                        ,
"_ncl_getProcessorNo"          ,
"_ncl_sendInterrupt"           ,
"_ncl_hostSync"                ,
"_ncl_hostSyncArray"           ,
"stop"                         ,
"_clock_initialize"            ,
"clock_initialize"             ,
"_clock"                       ,
"_Inter_Template"              ,
"_DefaultIntHandler"           ,
"_LibintIntManager"            ,
"_EnableInterrupts_PSWR"       ,
"_DisableInterrupts_PSWR"      ,
"_EnableInterrupts_IMR_Low"    ,
"_DisableInterrupts_IMR_Low"   ,
"_EnableInterrupts_IMR_High"   ,
"_DisableInterrupts_IMR_High"  ,
"_FlushInterrupts"             ,
"_GetInterruptMask_IMR_Low"    ,
"_GetInterruptMask_IMR_High"   ,
"_LibintGetCPUID"              ,
"_SetInterrupt"                ,
"_GetInterrupt"                ,
"_ClearInterrupt"              ,
"_LibintSetFPCCR"              ,
"_stopwatch_head"              ,
"_pr1intf"                      ,
"_puts_"                       ,
"_puts"                        ,
"FPIEEE_Signal"                ,
"_lstat"                       ,
"_utime"                       ,
"_fchmod"                      ,
"_fchown"                      ,
"_semi_fopen"                  ,
"_semi_fclose"                 ,
"_semi_fwrite"                 ,
"_semi_fread"                  ,
"_semi_fseek"                  ,
"__Z6dummy_v"				   ,
"__Z13beginMessage_v"			,
"__Z16completeMessage_15NM_IO_ServiceIDi",
"__main"						,
"__exit"

}




-----------------------------------------------------------------------------------
exclude_section_prefix[#exclude_section_prefix+1] = "____dummy____"
 
list = {}
func = {}
section = "null"
-- count = 0;
 
 
 
 local function contains(table, val)
	if #table==0 then 
		return true
	end
	for i=1,#table do
		--print(table[i],val);
		if table[i] == val then 
			return true
		end
	end
	return false
end
 
 local function contains_prefix(table, val)
   for i=1,#table do
	 --print(table[i],val);
	  indx,last=string.find(val,table[i]);
	  if (indx<>nul) then
		if indx==1 then
		     return true
        end
	  end
   end
   return false
end
 
local function file_exists(filename)
    local file = io.open(filename, "r")
    if (file) then
        -- Obviously close the file if it did successfully open.
        file:close()
        return true
    end
    return false
end 

local function space_wagon( n)
	local str=""
	for i = 1 , n do
		str=str.." ";
	end
	return str
end




print('//****************************************************************************************')
print('// This is automaticly generated asm-file by profiler-gcc.lua' );
print('//****************************************************************************************')
print('import from "d:/git/hal/include/nmprofiler.mlb";')
print('begin ".text_nmprofiler\"')
print("PROFILE_BEGIN("..max_funcname_length..");")
print('')
	
-- print(arg[1])

if #arg==1 then
if file_exists(arg[1]) then
	for line in io.lines(arg[1]) do 
		-- search of section ------------------------------------
		idx=string.find(line,"^%.text[%.%a_%d]*%s?");
		if idx<>nil then
			section=string.sub(line,string.find(line,"%.text[%.%a_%d]*"));
		end
		-- search of function ------------------------------------
		idx=string.find(line,"^%s+0x%x+%s+[_%a][_%a%d]+$");
		if idx<>nil then
			idx0,idx1=string.find(line,"0x%x+"); -- find address
			func.addr=string.sub(line,idx0,idx1); -- extract address
			func.name=string.sub(line, string.find(line,"[_%a%d]+",idx1+1)); -- extract name
			func.sect=section;
			
			-- count= count+1;
			if not contains_prefix(exclude_section_prefix, func.sect) then
				if contains(include_section, func.sect) then
					if not contains(exclude_name, func.name) then
						len = string.len(func.name);
						sss=space_wagon(max_funcname_length-len);
						print("PROFILE_FUNC("..func.name..","..sss.."\""..func.name..sss.."\");// "..func.addr.." ["..func.sect.."]");
						--print("\""..func.name.."\",");
						--print(count,func.addr,func.name,func.sect);
					end
				end
			end
		end
	end
end
end 
 
print('') 
print('PROFILE_END();')
print('end ".text_nmprofiler";')

File: script/profiler-gcc.py
import re
import os.path
import sys

program_name = sys.argv[0]
arguments = sys.argv[1:]
count = len(arguments)
mapfile =""
if count>0:
	#print(count,arguments)
	mapfile  = arguments[0]

section=""
max_funcname_length=16


#white list of section  RegExp patterns. Only matched sectinos will be passed
section_whitelist={'\.text_*'}

#black list of section RegExp patterns . Matched sections will be blocked
section_blacklist={
'\.text_nmprofiler',
'\.text_printf',
'\.text\.',
'\.text '
}

#black list of funcname RegExp patterns. Matched functions will be blocked
function_blacklist={
'ConvI32toF'	,
'ConvU32toF'	,
'ConvFtoI32'	,
'ConvFtoU32'	,
'ConvFtoI64'	,
'ConvFtoU64'	,
'ConvI32toD'	,
'ConvU32toD'	,
'FCmp'	,
'ConvDtoI32'	,
'ConvDtoU32'	,
'ConvDtoF'	,
'ConvDtoI64'	,
'ConvDtoU64'	,
'ConvI64toD'	,
'ConvU64toD'	,
'ConvFtoD'	,
'LShift64'	,
'RShift64'	,
'ARShift64'	,
'DCmp'	,
'DAdd'	,
'Mul64'	,
'DMul'	,
'IDiv64'	,
'UDiv64'	,
'UMod64'	,
'IMod64'	,
'DDiv'	,
'Mul32HighU'	,
'Mul32HighS'	,
'^vec_',
'__[a-zA-W]',
'_print_32x',
'_result_code'      ,
'loadStackAddr'     ,
'^start$'           ,
'^stop$'            ,
'^_sync$'           ,
'^_clock$'          ,
'^_exit$'           ,
'^_ncl'			    ,
'_clock_initialize' ,
'clock_initialize'  ,
'_\w*Interrupt'     ,
'_Libint'           ,
'_puts_'            ,
'_puts'             ,
'FPIEEE_Signal'     ,
'^_lstat'           ,
'^_utime'           ,
'^_fchmod'          ,
'^_fchown'          ,
'^_semi_'           ,
'^DResult$'			,
'^_stopwatch'		,
'^_Inter_Template$' ,
'^_DefaultIntHandler$'
}

nonstd_names = {
'UDiv32'    ,
'IDiv32'    ,
'LShift32'  ,
'ARShift32' ,
'UMod32'    ,
'IMod32'    ,
'Mul32'     ,
'RShift32'  ,
'FAdd'      ,
'FMul'		,    
'FDiv'		,
'ConvI64toF',
'ConvU64toF',
'Mul32Ex'   ,
'FFrExp'    ,
'DFrExp'    
#generateCall42
#'DAdd',
#'Mul64',
#'DMul',
#'IDiv64',
#'UDiv64',
#'UMod64',
#'IMod64',
#'DDiv'

}

print('//****************************************************************************************')
print('// This is automaticly generated asm-file by profiler-gcc.py' );
print('//****************************************************************************************')
print('import from "nmprofiler.mlb";')
print('begin ".text_nmprofiler\"')
print("PROFILE_BEGIN(",max_funcname_length,",0);")
print('')

address_list=[]
if os.path.isfile(mapfile):
	f = open(mapfile, 'r')
	
	for line in f:
		#try to detect pure section declaration without address
		match=re.search('^([.\w]+)\n',line)
		if match:
			section=match[1]
			continue
		else:
			#try to detect section declaration with address and size
			match=re.search('^\s?([.\w]+)\s+0x([\dabcdef]*)\s+0x([\dabcdef]*)',line)
			if match:
				section=match[1]
				continue
		
		
		#try to detect lebel(func or data) declaration with address 
		match=re.search('^\s+0x([\dabcdef]*)\s+([_a-zA-Z][_a-zA-Z0-9]+)',line)
		if match:
			funcname=match[2]
			funcaddr=match[1]
			#print('[',section,']', func)
		else:
			continue
		
		#filter out by section whiteslist 
		matched=False
		for white_pattern in section_whitelist:
			match=re.search(white_pattern,section)
			if match:
				matched=True
				break
		if not matched:
			continue
		
		#filter out by section blackslist 
		matched=False
		for black_pattern in section_blacklist:
			match=re.search(black_pattern,section)
			if match:
				matched=True
				continue
		if matched:
			continue
		
		#filter out by funcname blackslist 
		matched=False
		for black_pattern in function_blacklist:
			match=re.search(black_pattern,funcname)
			if match:
				matched=True
				continue
		if matched:
			continue
		

		#match on non-std C-call function
		matched=False
		for name in nonstd_names:
			if name==funcname:
				matched=True
				break

		#print(outfuncname)
		match=re.search("(_)?(_Z\d*)?(.*)",funcname)
		if match:
			outfuncname=match[3]
		
		match=re.search("(.+)PK.*",outfuncname)
		if match:
			outfuncname=match[1]
		
		match=re.search("(.+)RK.*",outfuncname)
		if match:
			#print("**********")
			outfuncname=match[1]
		
		
		align = ('                          ')[:16-len(outfuncname)]
		outfuncname= (outfuncname+" "*max_funcname_length)[:max_funcname_length]

		if funcaddr in address_list:
			print("\t//FUNCTION(",funcname,",",align,"\""+outfuncname+"\");// ",funcaddr," [",section,"]")
			continue
		address_list.append(funcaddr)

		if matched:
			print("\tNONCFUNC(",funcname,",",align,"\""+outfuncname+"\");// ",funcaddr," [",section,"]")
		else:				
			print("\tFUNCTION(",funcname,",",align,"\""+outfuncname+"\");// ",funcaddr," [",section,"]")
	


	
print('') 
print('PROFILE_END();')
print('end ".text_nmprofiler";')
	
